File: src/test/java/conservative/ExchangeEngineTest.java
package conservative;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.List;

public class ExchangeEngineTest {

    @Test
    public void shouldThrowWhenInputIsZero() {
        ExchangeEngine engine = new ExchangeEngine("USD");
        Assertions.assertThrows(AssertionError.class, () ->
           engine.exchange(0)
        );
    }

    @Test
    public void shouldThrowWhenInputIsNegative() {
        ExchangeEngine engine = new ExchangeEngine("USD");
        Assertions.assertThrows(AssertionError.class, () ->
            engine.exchange(-1)
        );
    }

    @Test
    public void shouldExchange_USD_87p_intoTheExpectedAmount() {
        ExchangeEngine engine = new ExchangeEngine("USD");
        List<Models.Denomination> expected = List.of(
                Models.Dollar.QUARTER_COIN,
                Models.Dollar.QUARTER_COIN,
                Models.Dollar.QUARTER_COIN,
                Models.Dollar.DIME_COIN,
                Models.Dollar.PENNY_COIN,
                Models.Dollar.PENNY_COIN
        );
        List<Models.Denomination> actual = engine.exchange(87);
        Assertions.assertArrayEquals(expected.toArray(), actual.toArray());
    }

    @Test
    public void shouldExchange_USD_287p_intoTheExpectedAmount() {
        ExchangeEngine engine = new ExchangeEngine("USD");
        List<Models.Denomination> expected = List.of(
                Models.Dollar.ONE_DOLLAR_NOTE,
                Models.Dollar.ONE_DOLLAR_NOTE,
                Models.Dollar.QUARTER_COIN,
                Models.Dollar.QUARTER_COIN,
                Models.Dollar.QUARTER_COIN,
                Models.Dollar.DIME_COIN,
                Models.Dollar.PENNY_COIN,
                Models.Dollar.PENNY_COIN
        );
        List<Models.Denomination> actual = engine.exchange(287);
        Assertions.assertArrayEquals(expected.toArray(), actual.toArray());
    }

    @Test
    public void shouldExchange_EUR_287ct_intoTheExpectedAmount() {
        ExchangeEngine engine = new ExchangeEngine("EUR");
        List<Models.Denomination> expected = List.of(
                Models.Euro.TWO_EURO_COIN,
                Models.Euro.FIFTY_CENT_COIN,
                Models.Euro.TWENTY_CENT_COIN,
                Models.Euro.TEN_CENT_COIN,
                Models.Euro.FIVE_CENT_COIN,
                Models.Euro.TWO_CENT_COIN);
        List<Models.Denomination> actual = engine.exchange(287);
        Assertions.assertArrayEquals(expected.toArray(), actual.toArray());
    }
}

File: src/test/kotlin/progressive/ExchangeEngineTest.kt
package progressive

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.lang.AssertionError
import kotlin.test.assertEquals

class ExchangeEngineTest {

    @Test
    fun `should throw when input is 0`() {
        val engine = ExchangeEngine("USD")
        assertThrows<AssertionError> {
            engine.exchange(0)
        }
    }

    @Test
    fun `should throw when input is negative`() {
        val engine = ExchangeEngine("USD")
        assertThrows<AssertionError> {
            engine.exchange(-1)
        }
    }

    @Test
    fun `test for USD 87 pennies`() {
        val engine = ExchangeEngine("USD")
        val expected = listOf(
            Dollar.QUARTER_COIN,
            Dollar.QUARTER_COIN,
            Dollar.QUARTER_COIN,
            Dollar.DIME_COIN,
            Dollar.PENNY_COIN,
            Dollar.PENNY_COIN)
        val actualDenominations = engine.exchange(87)
        assertEquals(expected, actualDenominations)
    }

    @Test
    fun `test for USD 287 pennies`() {
        val engine = ExchangeEngine("USD")
        val expected = listOf(
            Dollar.ONE_DOLLAR_NOTE,
            Dollar.ONE_DOLLAR_NOTE,
            Dollar.QUARTER_COIN,
            Dollar.QUARTER_COIN,
            Dollar.QUARTER_COIN,
            Dollar.DIME_COIN,
            Dollar.PENNY_COIN,
            Dollar.PENNY_COIN)
        val actualDenominations = engine.exchange(287)
        assertEquals(expected, actualDenominations)
    }

    @Test
    fun `test for EUR 287 cent`() {
        val engine = ExchangeEngine("EUR")
        val expected = listOf(
            Euro.TWO_EURO_COIN,
            Euro.FIFTY_CENT_COIN,
            Euro.TWENTY_CENT_COIN,
            Euro.TEN_CENT_COIN,
            Euro.FIVE_CENT_COIN,
            Euro.TWO_CENT_COIN)
        val actualDenominations = engine.exchange(287)
        assertEquals(expected, actualDenominations)
    }
}
